sdk/compiler/damlc/lib/DA/Cli/Damlc.hs

```diff
@@ -1471,38 +1471,48 @@ execClean :: ProjectOpts -> EnableMultiPackage -> MultiPackageLocation -> MultiP
-                  singleCleanEffect $ projectOpts {projectRoot = Just $ ProjectPath p}
-        (False, True) -> do
-          hPutStrLn stderr "Multi-package clean option used with multi-package disabled - re-enable it by setting the --enable-multi-package=yes flag."
-          exitFailure
-        _ -> singleCleanEffect projectOpts
+    effect
+      | getEnableMultiPackage enableMultiPackage
+      = do
+        mMultiPackagePath <- getMultiPackagePath multiPackageLocation
+        isProject <- withProjectRoot' (projectOpts {projectCheck = ProjectCheck "" False}) $ const $ doesFileExist projectConfigName
+        case (mMultiPackagePath, isProject, getMultiPackageCleanAll cleanAll) of
+          -- daml clean --all with no multi-package.yaml
+          (Nothing, _, True) -> do
+            hPutStrLn stderr "Couldn't find a multi-package.yaml at current or parent directory. Use --multi-package-path to specify its location."
+            exitFailure
+          -- daml clean --all with a multi-package.yaml
+          (Just path, _, True) ->
+            withMultiPackageConfig path $ \multiPackageConfig -> do
+              hPutStrLn stderr $ "Running multi-package clean of all packages in " <> unwrapProjectPath path
+              forM_ (mpPackagePaths multiPackageConfig) $ \p ->
+                singleCleanEffect $ projectOpts {projectRoot = Just $ ProjectPath p}
+          -- daml clean in a package
+          (_, True, False) -> singleCleanEffect projectOpts
+          -- daml clean outside a package and no multi-package.yaml
+          (Nothing, False, False) -> do
+            hPutStrLn stderr "No valid daml.yaml or multi-package.yaml could be found."
+            exitFailure
+          -- daml clean outside a package with a multi-package.yaml
+          (Just _, False, False) -> do
+            hPutStrLn stderr $ "No valid daml.yaml found, but a multi-package.yaml was found instead.\n"
+              <> "If you intended to clean everything within this multi-package.yaml, use the --all flag"
+            exitFailure
+      | getMultiPackageCleanAll cleanAll
+      = do
+        hPutStrLn stderr "Multi-package clean option used with multi-package disabled - re-enable it by setting the --enable-multi-package=yes flag."
+        exitFailure
+      | otherwise = singleCleanEffect projectOpts
 
 singleCleanEffect :: ProjectOpts -> IO ()
 singleCleanEffect projectOpts =
   withProjectRoot' projectOpts $ \_relativize -> do
     isProject <- doesFileExist projectConfigName
     when isProject $ do
-        let removeAndWarn path = do
-                whenM (doesDirectoryExist path) $ do
-                    putStrLn ("Removing directory " <> path)
-                    removePathForcibly path
-                whenM (doesFileExist path) $ do
-                    putStrLn ("Removing file " <> path)
-                    removePathForcibly path
-        removeAndWarn damlArtifactDir
-        putStrLn "Removed build artifacts."
```

Comment:
Probably good to keep the last line "Removed build artifacts" that confirms removal is complete.